manifests/install.pp

```diff
@@ -2,10 +2,38 @@
 #
 # This class is called from puppet_agent for install.
 #
-class puppet_agent::install {
+# === Parameters
+#
+# [package_file_name]
+#   The puppet-agent package file name.
+#   (see puppet_agent::prepare::package_file_name)
+#
+class puppet_agent::install(
+  $package_file_name = undef,
+) {
   assert_private()
 
-  package { $::puppet_agent::package_name:
-    ensure => present,
+  if $::operatingsystem == 'SLES' and $::operatingsystemmajrelease == '10' {
+    if $puppet_agent::params::master_agent_version != $::aio_agent_version {
+
+      contain puppet_agent::install::remove_packages
+
+      exec { 'replace puppet.conf removed by package removal':
```

Comment:
Solaris and AIX most likely - that would be @haus or @geoffnichols 

OSX - it does quasi upgrades. It will replace the package and all that good stuff, but it won't preserve any files. It will always over write them. However our puppet-agent package does that work for us of backing up puppet.conf and restoring it.  see https://tickets.puppetlabs.com/browse/RE-5089